// @unocss/preset-mini/dist/theme.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const colors = require('./shared/preset-mini.f3fc54d2.cjs');
const _default = require('./shared/preset-mini.b22ca5de.cjs');
require('./shared/preset-mini.78d4899a.cjs');
require('./shared/preset-mini.a21c5071.cjs');
require('@unocss/core');



exports.colors = colors.colors;
exports.baseSize = _default.baseSize;
exports.blur = _default.blur;
exports.borderRadius = _default.borderRadius;
exports.boxShadow = _default.boxShadow;
exports.breakpoints = _default.breakpoints;
exports.containers = _default.containers;
exports.dropShadow = _default.dropShadow;
exports.duration = _default.duration;
exports.easing = _default.easing;
exports.fontFamily = _default.fontFamily;
exports.fontSize = _default.fontSize;
exports.fontWeight = _default.fontWeight;
exports.height = _default.height;
exports.letterSpacing = _default.letterSpacing;
exports.lineHeight = _default.lineHeight;
exports.lineWidth = _default.lineWidth;
exports.maxHeight = _default.maxHeight;
exports.maxWidth = _default.maxWidth;
exports.preflightBase = _default.preflightBase;
exports.ringWidth = _default.ringWidth;
exports.spacing = _default.spacing;
exports.textIndent = _default.textIndent;
exports.textShadow = _default.textShadow;
exports.textStrokeWidth = _default.textStrokeWidth;
exports.theme = _default.theme;
exports.verticalBreakpoints = _default.verticalBreakpoints;
exports.width = _default.width;
exports.wordSpacing = _default.wordSpacing;
